File: Ahorn/entities/CeremonyFlame.jl
module CeremonyFlame

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/CeremonyFlame" CeremonyFlame(x::Integer, y::Integer, canSpread::Bool=true, isSource::Bool=true, maxSpreadDistance::Integer=200, spreadSpeed::Integer=20)

const placements = Ahorn.PlacementDict(
    "source" => Ahorn.EntityPlacement(CeremonyFlame),
    "static" => Ahorn.EntityPlacement(CeremonyFlame)
)

function Ahorn.selection(entity::CeremonyFlame)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::CeremonyFlame, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/ceremonyFlame.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local ceremonyFlame = {}

ceremonyFlame.name = "MaggyHelper/CeremonyFlame"
ceremonyFlame.depth = -50
ceremonyFlame.fieldInformation = {
    spreadSpeed = {
        minimumValue = 0
    },
    maxSpreadDistance = {
        minimumValue = 0
    }
}
ceremonyFlame.placements = {
    {
        name = "source",
        data = {
            isSource = true,
            spreadSpeed = 20,
            maxSpreadDistance = 200,
            canSpread = true
        }
    },
    {
        name = "static",
        data = {
            isSource = false,
            spreadSpeed = 20,
            maxSpreadDistance = 200,
            canSpread = false
        }
    }
}

function ceremonyFlame.sprite(room, entity)
    local state = entity.isSource and "burning" or "dormant"
    local sprite = drawableSprite.fromTexture("objects/ceremony_flame/" .. state, entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    sprite.color = {1.0, 0.4, 0.0, 0.9}
    return sprite
end

function ceremonyFlame.selection(room, entity)
    return utils.rectangle(entity.x - 12, entity.y - 48, 24, 48)
end

return ceremonyFlame
